client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -0,0 +1,477 @@
+		let sub_id_finalized = sub_id.clone();
+
+		let stream_finalized =
+			self.client.finality_notification_stream().map(move |notification| {
+				// We might not receive all new blocks reports, also pin the block here.
+				let _ = subscriptions.pin_block(&sub_id_finalized, notification.hash);
+
+				// The tree route contains the exclusive path from the latest finalized block
+				// to the block reported by the notification. Ensure the finalized block is
+				// properly reported to that path.
+				let mut finalized_block_hashes =
+					notification.tree_route.iter().cloned().collect::<Vec<_>>();
+				finalized_block_hashes.push(notification.hash);
+
+				FollowEvent::Finalized(Finalized {
+					finalized_block_hashes,
+					pruned_block_hashes: notification.stale_heads.iter().cloned().collect(),
+				})
+			});
+
+		let merged = tokio_stream::StreamExt::merge(stream_import, stream_finalized);
+
+		// The initialized event is the first one sent.
+		let finalized_block_hash = self.client.info().finalized_hash;
+		let finalized_block_runtime = generate_runtime_event(
+			&self.client,
+			runtime_updates,
+			&BlockId::Hash(finalized_block_hash),
+			None,
+		);
+
+		let _ = self.subscriptions.pin_block(&sub_id, finalized_block_hash);
+
+		let stream = stream::once(async move {
+			FollowEvent::Initialized(Initialized {
+				finalized_block_hash,
+				finalized_block_runtime,
+				runtime_updates,
+			})
+		})
+		.chain(merged);
+
+		let subscriptions = self.subscriptions.clone();
+		let fut = async move {
+			if let SubscriptionClosed::Failed(_) = sink.pipe_from_stream(stream.boxed()).await {
+				// The subscription failed to pipe from the stream.
+				let _ = sink.send(&FollowEvent::<Block::Hash>::Stop);
+			}
+
+			// The client disconnected or called the unsubscribe method.
+			subscriptions.remove_subscription(&sub_id);
+		};
+
+		self.executor.spawn("substrate-rpc-subscription", Some("rpc"), fut.boxed());
+		Ok(())
+	}
+
+	fn chain_head_unstable_body(
+		&self,
+		mut sink: SubscriptionSink,
+		follow_subscription: String,
+		hash: Block::Hash,
+		_network_config: Option<()>,
+	) -> SubscriptionResult {
+		let client = self.client.clone();
+		let subscriptions = self.subscriptions.clone();
+
+		let fut = async move {
+			let res = match subscriptions.contains(&follow_subscription, &hash) {
+				Err(SubscriptionError::InvalidBlock) => {
+					let _ = sink.reject(ChainHeadRpcError::InvalidBlock);
+					return
+				},
+				Err(SubscriptionError::InvalidSubId) => ChainHeadEvent::<String>::Disjoint,
+				Ok(()) => match client.block(&BlockId::Hash(hash)) {
+					Ok(Some(signed_block)) => {
+						let extrinsics = signed_block.block.extrinsics();
+						let result = format!("0x{}", HexDisplay::from(&extrinsics.encode()));
+						ChainHeadEvent::Done(ChainHeadResult { result })
+					},
+					Ok(None) => {
+						// The block's body was pruned. This subscription ID has become invalid.
+						ChainHeadEvent::<String>::Disjoint
```

Comment:
Do you send a `stop` notification somewhere?

Reply:
At the moment nope, will add a `stop` notification, thanks!